### ISSUE TITLE
Prefetching results in multiple parallel requests

### ISSUE THREAD

**Issue by mullender** [square/okhttp#1647]
When an application makes a prefetch request and, before the prefetch request completes, makes another request (or N requests) for the same resource. OkHttp will perform multiple identical requests in parallel. 
This ties up multiple executors, and downloads the same response twice, and causes the bandwidth to be shared, and increases server load. While not improving client side latency, which is the goal of prefetching.
This problem is especially bad on high latency or low bandwidth connections (mobile) when requesting larger resources (images).

The Volley library handles this case nicely:
https://android.googlesource.com/platform/frameworks/volley/+/master/src/main/java/com/android/volley/RequestQueue.java#61 (mWaitingRequests)
I propose to add a similar mechanism to Dispatcher: 
https://github.com/square/okhttp/blame/master/okhttp/src/main/java/com/squareup/okhttp/Dispatcher.java

I welcome feedback on this problem report, before I start on a PR.
@swankjesse you seem to be most familiar with Dispatcher

This is a unit test that demonstrates the problem 
```java
  @Test public void identicalRequestsShouldMerge() throws Exception {
    final CountDownLatch responseCountDownLatch = new CountDownLatch(7);
    final AtomicInteger serverResponseCount = new AtomicInteger(0);

    OkHttpClient client = new OkHttpClient();
    client.setCache(new Cache(new File("/tmp/"), 1024*1024*2));
    final MockWebServer server = new MockWebServer();
    server.start();
    server.setDispatcher(new com.squareup.okhttp.mockwebserver.Dispatcher() {
      @Override
      public MockResponse dispatch(final RecordedRequest request) throws InterruptedException {
        serverResponseCount.incrementAndGet();
        if (request.getPath().equals("/1")) {
          Thread.sleep(500, 0);
            return new MockResponse().setBody("delayed slow hello world"+request.getPath()).throttleBody(1, 100, TimeUnit.MILLISECONDS);
        }
          return new MockResponse().setBody("slow hello world"+request.getPath()).throttleBody(1, 100, TimeUnit.MILLISECONDS);
      }
    });
    dispatcher.setMaxRequests(5);
    dispatcher.setMaxRequestsPerHost(5);

    final Callback cb = new Callback() {
      @Override
      public void onFailure(final Request request, final IOException e) {
      }

      @Override
      public void onResponse(final Response response) throws IOException {
        logger.info(response.body().string()); // need to read the body, so it gets closes
        responseCountDownLatch.countDown();
      }
    };
    String url1 = server.getUrl("/1").toString();
    String url2 = server.getUrl("/2").toString();
    // create 5 parallel requests for url 1, in theory these should be merged into 1 request
    // and the response would be reused for all requests
    // in practice they consume all request executor threads, and block the requests for url 2 
    client.newCall(newRequest(url1)).enqueue(cb);
    client.newCall(newRequest(url1)).enqueue(cb);
    client.newCall(newRequest(url1)).enqueue(cb);
    client.newCall(newRequest(url1)).enqueue(cb);
    client.newCall(newRequest(url1)).enqueue(cb);

    // these only get executed after one of the requests for url 1 complete
    client.newCall(newRequest(url2)).enqueue(cb);
    client.newCall(newRequest(url2)).enqueue(cb);
    responseCountDownLatch.await();
    assertEquals("all requested should have responses", 0, responseCountDownLatch.getCount());
    assertEquals("only 2 requests should hit the server", 2, serverResponseCount.get());
  }
```

**Comment by JakeWharton** [comment]
This is essentially a dupe of #607.

**Comment by mullender** [comment]
Sorry about that, I did quite a few searches and did not come up with that one.

Does that mean that I should not work on a PR for version 2?

**Comment by JakeWharton** [comment]
Not necessarily. I think some of the Okio changes that have been made will make it easier to do on the multicasting of the response, but I'm not sure about the bits that actually cross into the underlying HTTP engine.

Also, we have this feature in Picasso which is also implemented in a class called `Dispatcher` so we draw inspiration from where as well.

**Comment by swankjesse** [comment]
I think it'll be quite difficult to do this without changing our current threading model. Right now each request is relatively independent, and that's simple.

If we join requests, we have to buffer data that's read by _call A_ but not yet by _call B_. When _call B_ catches up to _call A_, we don't want to block _call B_ just because _call A_ isn't consuming its response stream as quickly as possible. So we need either thread to be able to steal-the-lead from the other, which is weird and confusing.

Which is all to say, we're thinking a lot about this, and we're going to build it. But I want to build it right without just buffering everything in memory like Volley does.

**Comment by mullender** [comment]
Would it be possible to make the Dispatcher class non-final so people can use the setDispatcher call to inject their own dispatcher implementation?

I like how you worded the 'slow reader' problem. I haven't been able to create a nice solution for that yet. It requires a buffer for the content between the slowest reader's read position (A) and the fastest read position (B). In the worst case (A reads nothing) the whole response ends up in memory. Alternatively you can use a disk buffer, at the cost of performance and complexity.

**Comment by swankjesse** [comment]
@mullender I don't want to make Dispatcher non-final because I don't want to worry about API compatibility going forward. If you want to do your own dispatching, you can build your own dispatcher-like-thing on top of the synchronous API.

**Comment by swankjesse** [comment]
No action for us to take here, beyond the issues we're already tracking.
